### PR TITLE
GH-98831: Update generate_cases.py: register inst, opcode_metadata.h

### PR DESCRIPTION
**Pull request by gvanrossum**
(These aren't used yet, but may be coming soon,
and it's easier to keep this tool the same between branches.)

* Issue: gh-98831

### CODE REVIEW

File: Python/opcode_metadata.h [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    [CALL_FUNCTION_EX] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [MAKE_FUNCTION] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [RETURN_GENERATOR] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [BUILD_SLICE] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [FORMAT_VALUE] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [COPY] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [BINARY_OP] = { 2, 1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [SWAP] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [EXTENDED_ARG] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [CACHE] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
```
Comment (iritkatriel) [inline Python/opcode_metadata.h]: END_FOR is missing, I think we need something to handle macros. 

Also, note that missing entries default to 0. Would be safer if they defaulted to -1.

File: Python/opcode_metadata.h [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    [PRINT_EXPR] = { 1, 0, DIR_NONE, DIR_NONE, DIR_NONE },
+    [RAISE_VARARGS] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [INTERPRETER_EXIT] = { 1, 0, DIR_NONE, DIR_NONE, DIR_NONE },
+    [RETURN_VALUE] = { 1, 0, DIR_NONE, DIR_NONE, DIR_NONE },
+    [GET_AITER] = { 1, 1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [GET_ANEXT] = { 1, 2, DIR_NONE, DIR_NONE, DIR_NONE },
+    [GET_AWAITABLE] = { 1, 1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [SEND] = { -1, -1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [ASYNC_GEN_WRAP] = { 1, 1, DIR_NONE, DIR_NONE, DIR_NONE },
+    [YIELD_VALUE] = { 1, 0, DIR_NONE, DIR_NONE, DIR_NONE },
```
Comment (iritkatriel) [inline Python/opcode_metadata.h]: [stack_effect of YIELD_VALUE](https://github.com/python/cpython/blob/15aecf8dd70f82eb507d74fae9662072a377bdc8/Python/compile.c#L1128) is 0. Why is it here 1 popped, 0 pushed?

**Comment by iritkatriel** [comment]
This is the sanity check I used to find those two issues:

```
--- a/Python/compile.c
+++ b/Python/compile.c
@@ -8854,6 +8854,26 @@ assemble(struct compiler *c, int addNone)
 
     assert(no_redundant_jumps(g));
 
+    for (basicblock *b = g->g_entryblock; b != NULL; b = b->b_next) {
+        for (int j = 0; j < b->b_iused; j++) {
+            struct instr *instr = &b->b_instr[j];
+            int opcode = instr->i_opcode;
+            assert(opcode < 256);
+            int pushed = _PyOpcode_opcode_metadata[opcode].n_pushed;
+            int popped = _PyOpcode_opcode_metadata[opcode].n_popped;
+            if (pushed >= 0 && popped >= 0) {
+                int oparg = instr->i_oparg;
+                int effect = stack_effect(opcode, oparg, -1);
+                if (effect != pushed-popped) {
+                   fprintf(stderr, "opcode = %d  oparg = %d effect = %d pushed = %d popped = %d\n",
+                                opcode, oparg, effect, pushed, popped);
+                }
+                assert(effect == pushed-popped);
+            }
+        }
+    }
+
+

```

File: Python/opcode_metadata.h (hunk 1 above)
Comment (gvanrossum) [reply Python/opcode_metadata.h]: > Also, note that missing entries default to 0. Would be safer if they defaulted to -1.

The code generator doesn't know the numeric values of the opcodes (yet) so it doesn't have enough info to generate explicit entries filled with `-1` for missing values. The best we can do is add an additional flag `valid_entry`.

File: Python/opcode_metadata.h (hunk 2 above)
Comment (gvanrossum) [reply Python/opcode_metadata.h]: That seems to be a bug in bytecodes.c. The code ends with `goto resume_frame` so the push sequence at the end is never executed. I can fix it by making the IO effect `(retval -- retval)` (indicating that it doesn't change) or perhaps `(retval -- unused)`, indicating that we don't write it here. I presume it's written in some weird way when the frame is resumed.

**Comment by gvanrossum** [comment]
> This is the sanity check I used to find those two issues:
> [...]

Maybe we should include that in the code as a helper function and call it in an `#ifndef NDEBUG` block?

**Review by iritkatriel — APPROVED** [review]

**Comment by gvanrossum** [comment]
Oh shit. Will follow up with a fix.

**Comment by gvanrossum** [comment]
Have to revert, see https://github.com/python/cpython/pull/100777

**Comment by gvanrossum** [comment]
Bot failures fixed by https://github.com/python/cpython/pull/100778

**Pull request merged**